proxygen/lib/http/codec/compress/experimental/qcram/QCRAMHeader.h

```diff
@@ -0,0 +1,63 @@
+/*
+ *  Copyright (c) 2017, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#pragma once
+
+#include <proxygen/lib/http/codec/compress/HPACKHeader.h>
+
+#include <folly/Conv.h>
+#include <ostream>
+#include <string>
+#include <vector>
+
+namespace proxygen {
+
+class QCRAMHeader : public HPACKHeader {
+ public:
+  QCRAMHeader() {}
+
+  QCRAMHeader(folly::StringPiece name_,
+              folly::StringPiece value_):
+      HPACKHeader(name_, value_) {}
+
+  QCRAMHeader(QCRAMHeader&& goner) noexcept
+      : HPACKHeader(std::move(goner)),
+        epoch(goner.epoch),
+        refCount(goner.refCount),
+        dedupName(std::move(goner.dedupName)),
+        dedupValue(std::move(goner.dedupValue)) {}
+
+  QCRAMHeader& operator=(QCRAMHeader&& goner) noexcept {
+    std::swap(name, goner.name);
+    std::swap(value, goner.value);
+    epoch = goner.epoch;
+    refCount = goner.refCount;
+    std::swap(dedupName, goner.dedupName);
+    std::swap(dedupValue, goner.dedupValue);
+    return *this;
+  }
+
+  ~QCRAMHeader() {}
+
+  int32_t epoch{-1};
```

Comment:
I'm not sure epoch is used in a way that would work as QCRAM is currently specified.  The simulator is transferring the sequence numbers but they are not encoded on the wire anymore.  To use epoch in the current draft, we'd need to define it in terms of absolute index.  Since the table doesn't change size, HeaderTable can keep a wrap counter, and determine the absolute insert index of a current entry as wrapCount * size_ + i.  An encoder can track the decoder's highest ack'd index, and getIndex/nameIndex can use these two values to determine if a reference is vulnerable.  This would trade per-entry state for state per in-flight header block on the control stream.